Regression: Trybuild doesn't pass on `--target`, even when needed

We're using Trybuild in [LPC8xx HAL](https://github.com/lpc-rs/lpc8xx-hal). We have a [`.cargo/config.toml`](https://github.com/lpc-rs/lpc8xx-hal/blob/master/.cargo/config.toml) there that sets the target to `thumbv6m-none-eabi`.

We override the target to run the Trybuild tests using `--target` in our [build script](https://github.com/lpc-rs/lpc8xx-hal/blob/master/scripts/build.sh) (specifically [these lines](https://github.com/lpc-rs/lpc8xx-hal/blob/9f2661f6d8f6a918760e3dea8b296011938407d5/scripts/build.sh#L46-L49)). This works perfectly with Trybuild 1.0.43 (and earlier), but fails with Trybuild 1.0.44:
```
     Running `/home/hanno/Projects/lpc-rs/lpc8xx-hal/target/x86_64-unknown-linux-gnu/debug/deps/compiletest-a70653e23af66043`

running 1 test
    Checking lpc8xx-hal-tests v0.0.0 (/home/hanno/Projects/lpc-rs/lpc8xx-hal/target/tests/lpc8xx-hal)
error[E0463]: can't find crate for `std`
  |
  = note: the `thumbv6m-none-eabi` target may not support the standard library
  = note: `std` is required by `lpc8xx_hal_tests` because it does not declare `#![no_std]`

error: aborting due to previous error

For more information about this error, try `rustc --explain E0463`.
error: could not compile `lpc8xx-hal-tests`

To learn more, run the command again with --verbose.
test compile_test ... FAILED

failures:

---- compile_test stdout ----
thread 'compile_test' panicked at 'tests failed', /home/hanno/.cargo/registry/src/github.com-1ecc6299db9ec823/trybuild-1.0.44/src/run.rs:51:13
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace


failures:
    compile_test

test result: FAILED. 0 passed; 1 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.40s
```

This error indicates that `--target` didn't take, and that the target from `.cargo/config.toml` is used.

#121 seems to be the culprit here. I think the code it added wrongly assumes that Trybuild doesn't need to pass `--target`, if the `--target` it receives is the same as the host. This is not true, if `--target` is used to override a target set in `.cargo/config.toml`/`.cargo/config`.

cc @taiki-e, @dtolnay

--- a/build.rs
+++ b/build.rs
@@ -5,34 +5,7 @@ use std::path::Path;
 
 fn main() -> io::Result<()> {
     let out_dir = env::var_os("OUT_DIR").unwrap();
-    let mut target = env::var("TARGET").ok();
-
-    // When --target flag is passed, cargo does not pass RUSTFLAGS to rustc when
-    // building proc-macro and build script even if the host and target triples
-    // are the same. Therefore, if we always pass --target to cargo, tools such
-    // as coverage that require RUSTFLAGS do not work for tests run by trybuild.
-    //
-    // To avoid that problem, do not pass --target to cargo if we know that it
-    // has not been passed.
-    //
-    // Cargo does not have a way to tell the build script whether --target has
-    // been passed or not, so we use the following heuristic:
-    //
-    // - The host and target triples are the same.
-    // - And RUSTFLAGS is available when *building* the build script.
-    //
-    // Note that the second is when building, not when running. This is due to:
-    //
-    // - After rust-lang/cargo#9601, cargo does not pass RUSTFLAGS to the build
-    //   script when running.
-    // - CARGO_ENCODED_RUSTFLAGS, which was introduced in rust-lang/cargo#9601,
-    //   cannot be used for this purpose because it contains the value of
-    //   RUSTFLAGS even if --target is passed and the host and target triples
-    //   are the same.
-    if target == env::var("HOST").ok() && option_env!("RUSTFLAGS").is_some() {
-        target = None;
-    }
-
+    let target = env::var("TARGET").ok();
     let path = Path::new(&out_dir).join("target.rs");
     let value = match target {
         Some(target) => format!(r#"Some("{}")"#, target.escape_debug()),

--- a/build.rs
+++ b/build.rs
@@ -11,6 +11,6 @@ fn main() -> io::Result<()> {
         Some(target) => format!(r#"Some("{}")"#, target.escape_debug()),
         None => "None".to_owned(),
     };
-    let content = format!("const TARGET: Option<&'static str> = {};", value);
+    let content = format!("const TARGET: Option<&str> = {};", value);
     fs::write(path, content)
 }

--- a/src/lib.rs
+++ b/src/lib.rs
@@ -245,6 +245,22 @@ mod normalize;
 mod run;
 mod rustflags;
 
+// When --target flag is passed, cargo does not pass RUSTFLAGS to rustc when
+// building proc-macro and build script even if the host and target triples are
+// the same. Therefore, if we always pass --target to cargo, tools such as
+// coverage that require RUSTFLAGS do not work for tests run by trybuild.
+//
+// To avoid that problem, do not pass --target to cargo if we know that it has
+// not been passed.
+//
+// Currently, cargo does not have a way to tell the build script whether
+// --target has been passed or not, and there is no heuristic that can handle
+// this well.
+//
+// Therefore, expose a cfg to always treat the target as host.
+#[cfg(trybuild_no_target)]
+const TARGET: Option<&str> = None;
+#[cfg(not(trybuild_no_target))]
 include!(concat!(env!("OUT_DIR"), "/target.rs"));
 
 use std::cell::RefCell;

